### PR TITLE
feat(core): implement ɵgetInjectorMetadata debug API

### PR DESCRIPTION
**Pull request by AleksanderBodurri**
See commit

### CODE REVIEW

File: packages/core/src/render3/util/injector_discovery_utils.ts [hunk 1]
```diff
@@ -390,6 +391,37 @@ export function getInjectorProviders(injector: Injector): ProviderRecord[] {
   throwError('getInjectorProviders only supports NodeInjector and EnvironmentInjector');
 }
 
+export function getInjectorMetadata(injector: Injector): {type: string; name: string} {
```
Comment (AndrewKushnir) [inline packages/core/src/render3/util/injector_discovery_utils.ts:394]: Could you please add JSDoc to this function?

File: packages/core/src/render3/util/injector_discovery_utils.ts [hunk 2]
```diff
@@ -390,6 +391,37 @@ export function getInjectorProviders(injector: Injector): ProviderRecord[] {
...
+  let injectorMetadata;
+
+  if (injector instanceof NodeInjector) {
+    const lView = getNodeInjectorLView(injector);
+    const tNode = getNodeInjectorTNode(injector)!;
+
+    const directiveNames = [];
+    for (let i = tNode.directiveStart; i < tNode.directiveEnd; i++) {
+      assertDefined(
+          lView[i].constructor, 'Could not determine directive name on element injector.');
```
Comment (AndrewKushnir) [inline packages/core/src/render3/util/injector_discovery_utils.ts:404]: I'm wondering if we should return `null` in cases when we can not collect metadata. What's the most common behavior in other `ng.*` functions in similar cases?

File: packages/core/src/render3/util/injector_discovery_utils.ts [hunk 3]
```diff
@@ -390,6 +392,43 @@ export function getInjectorProviders(injector: Injector): ProviderRecord[] {
...
+    const lView = getNodeInjectorLView(injector);
+    const tNode = getNodeInjectorTNode(injector)!;
+    assertTNodeForLView(tNode, lView);
+    assertDefined(lView[tNode.index][HOST], 'Could not find node in element view.');
+
+    return {type: 'Element', source: lView[tNode.index][HOST]};
+  }
+
+  if (injector instanceof R3Injector && injector.source !== null) {
+    return {type: 'Environment', source: injector.source};
```
Comment (AndrewKushnir) [inline packages/core/src/render3/util/injector_discovery_utils.ts:422]: I think it's pretty common to not have a source info on R3Injector. Do we still want to return a metadata object in this case to indicate that it's an environment injector? 
```suggestion
  if (injector instanceof R3Injector) {
    return {type: 'Environment', source: injector.source ?? null};
```

Comment (AleksanderBodurri) [reply packages/core/src/render3/util/injector_discovery_utils.ts]: Good point, added

File: packages/core/src/render3/util/injector_discovery_utils.ts [hunk 4]
```diff
@@ -390,6 +392,43 @@ export function getInjectorProviders(injector: Injector): ProviderRecord[] {
...
+    return {type: 'Element', source: lView[tNode.index][HOST]};
+  }
+
+  if (injector instanceof R3Injector) {
+    return {type: 'Environment', source: injector.source ?? null};
+  }
+
+  if (injector instanceof NullInjector) {
+    return {type: 'Null', source: 'Null Injector'};
+  }
```
Comment (AndrewKushnir) [inline packages/core/src/render3/util/injector_discovery_utils.ts:427]: I'd propose lowercasing the type string and using `null` as a source for the NullInjector.
```suggestion
    return {type: 'element', source: lView[tNode.index][HOST]};
  }

  if (injector instanceof R3Injector) {
    return {type: 'environment', source: injector.source ?? null};
  }

  if (injector instanceof NullInjector) {
    return {type: 'null', source: null};
  }
```
The conversion of metadata -> strings for UI can be done on the API consumer side (e.g. in DevTools code).

**Review by AndrewKushnir — APPROVED** [review]
👍

**Comment by alxhub** [comment]
This PR was merged into the repository by commit a54713c8316787eea160cfdb7f2778a087fe59ed.

**Pull request closed without merging**